### PR TITLE
wg_engine: fix strokes triangulation artifacts

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
Fixed visual arifacts:

* angular circles
* extra triangles
* wrongly closed polylines

Animations to verify:
* 5344-honey-sack-hud.json
* birth_stone_logo.json
* insta_camera.json
* polystar.json
* starstrips.json

see: https://github.com/thorvg/thorvg/issues/2685

### CODE REVIEW

File: src/renderer/wg_engine/tvgWgGeometry.cpp [hunk 1]
```diff
@@ -65,7 +65,7 @@ void WgPolyline::appendPoint(WgPoint pt)
 {
     if (pts.count > 0) {
         float distance = pts.last().dist(pt);
-        if (distance > 0) {
+        if (!tvg::zero(distance * 1e-1)) {
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgGeometry.cpp:68]: can you please explain why tvg::zero is not enough?

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgGeometry.cpp:68]: also you are using different factors in different places for values of different meanings. can you also explain please?

Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgGeometry.cpp:68]: @mgrudzinska 
In the animations sometimes we can meet shapes with polylines like this:
```
    auto shape = tvg::Shape::gen();
    shape->moveTo(-1.74846e-06, -40);
    shape->lineTo(34.641, 20);
    shape->lineTo(-34.641, 20);
    shape->lineTo(4.76995e-07, -40);
    shape->close();
```
They are manually closed, but not close enough to detect it.

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgGeometry.cpp:68]: I don't think this is the way to solve this. 
try this:
```
        auto shape = tvg::Shape::gen();
        shape->moveTo(-1e-05, -40);         //try also: shape->moveTo(-1.1e-05, -40); 
        shape->lineTo(34.641, 20);
        shape->lineTo(-34.641, 20);
        shape->lineTo(0, -40);
        shape->close();
        shape->strokeFill(255, 0, 0);
        shape->strokeWidth(15);
        shape->translate(200,200);
        canvas->push(std::move(shape));
```
the result is clearly not perfect for SW. and GL (but in exaclty the same way not perfect), but it works for WG. if we change this shape to start in ```shape->moveTo(-1.1e-05, -40); ``` then the results for SW and GL don't change (still not perfect in the same way), but in WG there is a different problem. 

I think that if such cases are present in  examples we have to consider changing the epsilon value, but for all engines in the same way. 
I don't understand why WG needs a different value of epsilon especially only in one place.

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgGeometry.cpp:68]: you are right (private correspondence) that non of the engines rely on the epsilon values used in tvg::zero. 

I think that you may need you own epsilon, but let me test some cases and I'll come back to this

Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgGeometry.cpp]: Good idea to add own epsilon. 
Please, check the solution

File: src/renderer/wg_engine/tvgWgRenderData.cpp [hunk 2]
```diff
@@ -375,21 +377,21 @@ void WgRenderDataShape::updateMeshes(WgContext &context, const RenderShape &rsha
         if (!rshape.stroke->strokeTrim(trimBegin, trimEnd)) { trimBegin = 0.0f; trimEnd = 1.0f; }
         if (rshape.stroke->trim.simultaneous) {
             for (uint32_t i = 0; i < polylines.count; i++)
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgRenderData.cpp:379]: this is not correct. please check insta_camera.json - rounded corners are not applied

**Review by mgrudzinska — DISMISSED** [review]
please add some description in commit msg, especially since your are fixing more than one problem here. But also would be great, if you could split this commit into two separate commits. now one of the fixes is ok, but the other one is not, so the first one can not be applied.

File: src/renderer/wg_engine/tvgWgGeometry.cpp (hunk 1 above)
Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgGeometry.cpp]: now I like it!

and I found the problem, which was bothering me, why the result is not symmetrical - https://github.com/thorvg/thorvg/pull/2743

File: src/renderer/wg_engine/tvgWgRenderData.cpp (hunk 2 above)
Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:379]: this comment should be next to l.354...

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:379]: it is still a valid comment.
Lottie example, 1280x1280, NUM_PER_ROW 3, NUM_PER_COL 2:

current:
Zrzut ekranu 2024-09-18 o 01 28 01

expected:
Zrzut ekranu 2024-09-18 o 01 28 10

**Comment by mgrudzinska** [comment]
@SergeyLebedkin please try to keep the convention in commit msg:
```
1 line summary

description 
```
please note the empty line between summary and description

**Review by mgrudzinska — APPROVED** [review]

File: src/renderer/wg_engine/tvgWgGeometry.h [hunk 3]
```diff
@@ -28,6 +28,8 @@
 #include "tvgRender.h"
 #include "tvgArray.h"
 
+#define WG_FLOAT_EPSILON 1e-4
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgGeometry.h:31]: --

File: src/renderer/wg_engine/tvgWgGeometry.h [hunk 4]
```diff
@@ -70,6 +72,8 @@ struct WgMath
 
     void initialize();
     void release();
+
+    static bool zero(float val) { return (fabsf(val) <= WG_FLOAT_EPSILON) ? true : false; }
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgGeometry.h:76]: --

File: src/renderer/wg_engine/tvgWgGeometry.cpp [hunk 5]
```diff
@@ -65,7 +65,7 @@ void WgPolyline::appendPoint(WgPoint pt)
 {
     if (pts.count > 0) {
         float distance = pts.last().dist(pt);
-        if (distance > 0) {
+        if (!WgMath::zero(distance)) {
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgGeometry.cpp:68]: @SergeyLebedkin Please keep your original proposal:
` if (!tvg::zero(distance * 1e-1)) {`

If possible, you can add a comment with this special handling 'reason' there.

File: src/renderer/wg_engine/tvgWgGeometry.cpp [hunk 6]
```diff
@@ -334,12 +334,12 @@ void WgGeometryData::appendStrokeJoin(const WgPoint& v0, const WgPoint& v1, cons
...
     WgPoint offset1 = nrm1 * halfWidth;
     if (join == StrokeJoin::Round) {
-        appendCircle(v1, halfWidth);
+        appendCircle(v1, halfWidth, scale);
     } else if (join == StrokeJoin::Bevel) {
         appendRect(v1 - offset0, v1 + offset0, v1 - offset1, v1 + offset1);
     } else if (join == StrokeJoin::Miter) {
         WgPoint nrm = (nrm0 + nrm1);
-        if (!tvg::zero(dir0.x * dir1.y -  dir0.y * dir1.x)) {
+        if (!WgMath::zero(dir0.x * dir1.y - dir0.y * dir1.x)) {
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgGeometry.cpp:342]: Please revert to `tvg::zero()`

**Review by hermet — COMMENTED** [review]
@SergeyLebedkin Please check comments, it is deemed unnecessary because you are trying to hide an root problem with a fake zero. I think we should keep use one unified math function in ThorVG.  It is not good to keep spreading code fragments like this: https://github.com/thorvg/thorvg/issues/2313

**Review by hermet — APPROVED** [review]

**Pull request merged**